File: src/bridge/bridge.js
import { spawn } from "child_process";

const PROTOCOL_VERSION = 1;

export function createBridge(proc) {
  let requestId = 0;
  const pending = new Map();
  let buffer = "";

  proc.stdout.on("data", (data) => {
    buffer += data.toString();
    const lines = buffer.split("\n");
    buffer = lines.pop();

    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const envelope = JSON.parse(line);

        if (envelope.protocol_version && envelope.protocol_version !== PROTOCOL_VERSION) {
          process.stderr.write(`[bridge] protocol version mismatch: got v${envelope.protocol_version}, expected v${PROTOCOL_VERSION}\n`);
        }

        const response = envelope.payload || envelope;
        const id = envelope.request_id || response._request_id || response.request_id;
        if (id && pending.has(id)) {
          const { resolve } = pending.get(id);
          pending.delete(id);
          resolve({ ...response, _meta: { protocol_version: envelope.protocol_version } });
        }
      } catch {}
    }
  });

  return {
    async request(payload) {
      const id = ++requestId;
      return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
          pending.delete(id);
          resolve({ status: "error", error: "request timed out" });
        }, 30000);

        pending.set(id, {
          resolve: (resp) => {
            clearTimeout(timeout);
            resolve(resp);
          },
        });

        const envelope = {
          protocol_version: PROTOCOL_VERSION,
          request_id: id,
          type: payload.type,
          payload: payload,
        };

        const msg = JSON.stringify(envelope);
        proc.stdin.write(msg + "\n");
      });
    },

    close() {
      try {
        proc.stdin.end();
      } catch {}
    },
  };
}

File: engine/bridge.py
import sys
import json

from engine.errors import BridgeError

PROTOCOL_VERSION = 1


class Bridge:
    def __init__(self, stdin, stdout):
        self.stdin = stdin
        self.stdout = stdout

    def read_request(self):
        line = self.stdin.readline()
        if not line:
            return None
        line = line.strip()
        if not line:
            return None
        try:
            request = json.loads(line)
        except json.JSONDecodeError:
            raise BridgeError("invalid JSON in request", direction="in")

        if "protocol_version" not in request:
            self.log("warning: request missing protocol_version, assuming v1")
            request["protocol_version"] = 1

        if request["protocol_version"] != PROTOCOL_VERSION:
            self.log(
                f"warning: protocol version mismatch: got {request['protocol_version']}, expected {PROTOCOL_VERSION}"
            )

        return request

    def write_response(self, response, request=None):
        envelope = {
            "protocol_version": PROTOCOL_VERSION,
            "payload": response,
        }

        if request:
            if "request_id" in request:
                envelope["request_id"] = request["request_id"]
            elif "_request_id" in request:
                envelope["request_id"] = request["_request_id"]

        data = json.dumps(envelope)
        self.stdout.write(data + "\n")
        self.stdout.flush()

    def log(self, message):
        print(f"[ENGINE] {message}", file=sys.stderr, flush=True)

    def close(self):
        pass
